tests/parameter_flash_storage.cpp

```diff
@@ -0,0 +1,356 @@
+#include <CppUTest/TestHarness.h>
+#include <CppUTestExt/MockSupport.h>
+#include "../parameter_flash_storage.h"
+#include "../parameter_flash_storage_private.h"
+#include <parameter/parameter_msgpack.h>
+#include <cstdio>
+#include <cstring>
+#include "crc/crc32.h"
+
+TEST_GROUP(ConfigSaveTestCase)
+{
+    uint8_t data[128];
+    parameter_namespace_t ns;
+
+    void setup()
+    {
+        mock("flash").ignoreOtherCalls();
+        parameter_namespace_declare(&ns, NULL, NULL);
+    }
+};
+
+TEST(ConfigSaveTestCase, SavingConfigErasesPages)
+{
+    // In case we cannot find a valid block we erase the sector, just in case
+    mock("flash").expectOneCall("erase").withParameter("sector", data);
+
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+}
+
+TEST(ConfigSaveTestCase, SavingConfigLockUnlock)
+{
+    mock("flash").strictOrder();
+
+    mock("flash").expectOneCall("unlock");
+    mock("flash").expectOneCall("erase").withParameter("sector", data);
+    mock("flash").expectOneCall("write"); // data
+    mock("flash").expectOneCall("write"); // CRC(len)
+    mock("flash").expectOneCall("write"); // len
+    mock("flash").expectOneCall("write"); // crc(data)
+    mock("flash").expectOneCall("lock");
+
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+
+    // We need to manually check expectations since we tuned the flash mock to preserve ordering.
+    mock("flash").checkExpectations();
+}
+
+static void err_cb(void *p, const char *id, const char *err)
+{
+    (void) p;
+    (void) id;
+    (void) err;
+    char msg[128];
+    snprintf(msg, sizeof msg, "MessagePack error on item \"%s\": \"%s\"", id, err);
+
+    FAIL(msg);
+}
+
+TEST(ConfigSaveTestCase, SavingConfigWorks)
+{
+    // Declare a parameter and gives it a value.
+    parameter_t p;
+    parameter_integer_declare(&p, &ns, "foo");
+    parameter_integer_set(&p, 10);
+
+    // Check that the flash writer is used
+    // Number of expected written bytes is implementation-dependent
+    // Change if if necessary
+    for (int i = 0; i < 8; i++) {
+        mock("flash").expectOneCall("write");
+    }
+
+    // Saves the parameter, then change its value
+    parameter_flash_storage_save(data, sizeof(data), &ns);
+    parameter_integer_set(&p, 20);
+    CHECK_EQUAL(20, parameter_integer_get(parameter_find(&ns, "/foo")));
+
+    // Loads the config back from saved state
+    // We add an offset to skip the CRC and the block length
+    parameter_msgpack_read(&ns,
+                           (char *)(&data[PARAMETER_FLASH_STORAGE_HEADER_SIZE]),
+                           sizeof(data) - PARAMETER_FLASH_STORAGE_HEADER_SIZE,
+                           err_cb, NULL);
+
+    // Check that the parameter has the same value as saved
+    CHECK_EQUAL(10, parameter_integer_get(parameter_find(&ns, "/foo")));
```

Comment:
[suggestion] I would split this in two tests: one testing the write, one testing the read. So that there is one check per test (single point of failure).
You could factor out the parameter declaration and writing into a helper method inside the fixture/testcase.

Reply:
Actually there is only one code under test: the flash storage writer. The read is done using the parameter_msgpack API, which has its own tests. Or am I missing something ? The goal of this test is to make sure the written data structure is valid msgpack.

Reply:
So the first CHECK_EQUAL is only verifying that the parameter is set correctly, makes sense. While the second CHECK_EQUAL is actually checking that parameter_msgpack_read loads the tree from flash. Maybe abstracting away part of the test boiler plate could help, otherwise leave it like that.